python/tvm/relay/frontend/tensorflow.py

```diff
@@ -2029,6 +2096,73 @@ def if_node(self):
         return self._if
 
 
+class LoopBound(ExprVisitor):
+    """
+    When a loop body is create, we get a Relay expression backtracing all
+    the way back to input node. This will result in lots of unnecessary
+    expression placed into loop body and compute multiple times. For example,
+    consider the following tensorflow code:
+
+    .. code-block:: python
+
+        i = tf.constant(0)
+        data = tf.compat.v1.placeholder(tf.float32, shape=(1024, 1024))
+        slice = tf.strided_slice(data, 0, 512)
+        def c(i): return tf.less(i, 10)
+        def b(i): return [tf.add(i, 1), tf.add(i, 1) + slice]
+        r = tf.while_loop(c, b, [i])
+
+    If we directly create recursive function, slice will be placed into function body.
```

Comment:
Maybe we need to add a loop-invariant code motion (LICM) pass so that we don't need to handle this in different frameworks. It can benefit general Relay program as well, though it is not necessarily in this PR.

Reply:
Yes. In the future we can add a pass automatically detect invariant part inside loop/function, I'll add a TODO here.